Fix `versionadded` directive formatting

tsbinns opened the pull request:
#### What does this implement/fix?

Noticed some of the `versionadded` Sphinx directives were not rendering properly in the docs. Checked also for other directives while I was at it, but I only saw this with cases of `versionadded`.

tsbinns commented:
@larsoner `source-read` gives access to the content in `autoexamples` and `autotutorials` which we can check for bad formatting, but the rst files in `generated` don't contain the explicit docstrings (just `autoclass`/`autofunction`).

Is there a way to access the docstrings at the read phase, or do we have to rely on e.g., checking the html files once they've been written?

tsbinns commented:
Sorry for the delay, had to fight some new sphinx-build - Windows compatibility issues I did not have before.

> How about one of
> https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#docstring-preprocessing

That was exactly it, thanks @larsoner!

Have updated the script so it'll also check docstrings. Couple things to note which I will look into later:
- The code could probably be tidied up.
- There are some cases where warnings are repeated, but I'm not sure why.
- I also have some concerns about efficiency, since in this state it's checking the contents of all sources line-by-line, but I haven't tested to see how much this is slowing the docs building down.

tsbinns commented:
It's also picking up this case in the examples where the references aren't properly separated from the code.

https://github.com/mne-tools/mne-python/blob/84bc971d1bb7cb330845768d23f71ee31a1f7390/examples/preprocessing/css.py#L94-L105

<img width="877" height="851" alt="image" src="https://github.com/user-attachments/assets/d805b25d-90f3-4fe2-a7fa-72a617bba99d" />

This gives the following warning: `WARNING: File 'auto_examples/preprocessing/css' is missing a blank line before the directive '.. footbibliography::'`. So the message doesn't work great since there is a blank line, it's just not processed as such.
Is it good enough that the warning is pointing to something being wrong here, or should there be a different warning message for this special case?

larsoner commented:
> Is it good enough that the warning is pointing to something being wrong here, or should there be a different warning message for this special case?

I think it's okay if the warning isn't perfect / correct, as long as there is something to actually fix (and there is here!). What would be good to avoid is false alarms

tsbinns commented:
> What would be good to avoid is false alarms

Everything that gets picked up now is a genuine case of the directive not rendering properly.

> and if there are some false alarms, we can build some mechanism / allowlist of exceptions.

The exceptions I found so far where a blank line is not necessary are:
- it's right after a header marked by a line of all `----`, `====`, or `^^^^` (e.g., in examples, tutorials, docstring notes)
- the previous line is a directive or is content for a directive

https://github.com/mne-tools/mne-python/blob/84bc971d1bb7cb330845768d23f71ee31a1f7390/doc/sphinxext/directive_formatting.py#L46-L66

tsbinns commented:
Just pushed a commit that:
- expanded the comments.
- added a check that will catch a change in the number of args from the Sphinx events if that ever happens in a future version.
- added an initial check that a directive occurs anywhere in the whole file/docstring before searching line-by-line.

Warnings are still the same (https://app.circleci.com/pipelines/github/mne-tools/mne-python/29136/workflows/b74c3e86-2455-40c9-ba7b-384d5796bb65/jobs/76730/parallel-runs/0/steps/0-143), will fix these cases now.
